File: CP/LeetCode/C/153.c
// 153. Find Minimum in Rotated Sorted Array

#include <stdio.h>

int findMin(int *nums, int numsSize)
{
    int low = 0;
    int high = numsSize - 1;

    while (low < high)
    {
        int mid = (high + low) / 2;

        if (nums[mid] > nums[high])
        {
            // The minimum is in the right half
            low = mid + 1;
        }
        else
        {
            // The minimum is in the left half or at mid
            high = mid;
        }
    }
    return nums[low];
}

int main()
{
    int nums[] = {3, 4, 5, 1, 2};
    int numsSize = 5;

    int min = findMin(nums, numsSize);
    printf("The minimum in the rotated sorted array is: %d\n", min);

    return 0;
}

File: CP/LeetCode/Python/153.py
# 153. Find Minimum in Rotated Sorted Array

class Solution:
    def findMin(self, nums: list[int]) -> int:
        """
        Find the minimum element in a rotated sorted array.
        """
        left, right = 0, len(nums) - 1
        
        while left < right:
            mid = (left + right) // 2
            
            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid
        
        return nums[left]
    
if __name__ == "__main__":
    solution = Solution()
    nums = [3, 4, 5, 1, 2]
    
    print(solution.findMin(nums))  # Output: 1